core/src/main/java/org/apache/calcite/prepare/RelOptTableImpl.java

```diff
@@ -118,9 +115,9 @@ public static RelOptTableImpl create(
       RelDataType rowType,
       List<String> names,
       Table table,
-      Expression expression) {
+      Function<Class, Expression> expressionFunction) {
     return new RelOptTableImpl(schema, rowType, names, table,
-        c -> expression, table.getStatistic().getRowCount());
+        expressionFunction, table.getStatistic().getRowCount());
```

Comment:
`expressionFunction` is not a very good abstraction. That didn't matter when it was private. But now it is public. How can we improve the abstraction?

Reply:
Agree. I've introduced the `TableExpressionFactory` interface and use it instead of the previous `expressionFunction`.